src/main/java/org/bytedeco/javacpp/tools/Parser.java

```diff
@@ -3660,7 +3669,16 @@ boolean group(Context context, DeclarationList declList) throws ParserException
         } else if (info != null && info.flatten) {
             info.javaText = decl.text;
         }
-        declList.add(decl);
+        if (declList.add(decl)) {
+            for (Declaration dd: declList2.nonMemberDeclarations) {
+                // Only add the friend functions that we know will be visible through ADL.
+                for (int i=0; i<dd.declarator.parameters.declarators.length; i++)
+                    if (dd.declarator.parameters.declarators[i].type.cppName.equals(type.cppName)) {
+                        globalDeclList.add(dd);
```

Comment:
Am I missing something or can this not be done without modifying DeclarationList? It seems like we should be able to check for type.friend here and there, resetting as necessary to pass it as a normal static function, and do what's needed without having to keep a list of "nonMemberDeclarations".

Reply:
Also, why do we need to put this in the "global" list? We can put static methods anywhere in any class...

Reply:
The `nonMemberDeclarations` list is here to postpone the addition of the friend functions after the addition of the whole group AND if the addition of the group worked (`declList.add` returned `true`). Two reasons for this : 
1. if we directly add the friend functions when they are parsed, we end up with template types not replaced by java types. Apparently it's only when the whole group can be successfully added that we have the context info needed to replace the template types. Maybe there is a better way, I'm far from mastering how the parser handles templates.
2. if we add the friend function before the group to the global decl list, it will follow the class comment declaration, if any, and the comment will be attached to the friend function instead of correcty copied to the class file.

Concerning whether to translate the friend functions as static function of the global class or of the class they are friend of: both are possible, but I found more logical to make them methods of the global class. For example a function or operator combining 2  instances of 2 classes could be friend of both. Also many friend functions are also declared outside the class, to allow normal lookup instead of ADL only, in this case they end up in the global class.  There are already some equals, notEquals, or shiftLeft operators in the global class for this reason. Generally speaking friend functions are global functions, not member of the class. They appear in the class body only to be able to access private and protected members. Finally the user will often make a static import of the global class, allowing to not prefix the method like in `equals(iterator, end)`, instead of `named_module_iterator.equals(iterator, end)`.

Reply:
Well, to make this more consistent, I think friend functions should appear in the class where they are declared. If they are also supposed to be part of the global namespace, they will appear as such as normal non-friend static functions, too, right?

Reply:
Besides, it makes it easier to pair them with non-static methods like this, and also makes it clearer to users that these two methods are related:
```java
public class module_iterator extends Pointer {
...
@Namespace public static native @Name("operator !=") boolean notEquals(@Const @ByRef module_iterator a, @Const @ByRef module_iterator b);
@Override public boolean equals(Object o) { return o != null && o instanceof module_iterator && !notEquals(this, o); }
...
}
```
I'm not saying you need to do that right away, but if we start moving function declarations around, it makes it harder to bring them back later like this, and that's what needs to be done anyway because that's usually the intent of the original C++ code when declaring friends functions like that.

Reply:
From cppreference.com: 

> Names introduced by [friend](https://en.cppreference.com/w/cpp/language/friend) declarations within a non-local class X become members of the innermost enclosing namespace of X, but they do not become visible to ordinary [name lookup](https://en.cppreference.com/w/cpp/language/lookup) (neither [unqualified](https://en.cppreference.com/w/cpp/language/unqualified_lookup) nor [qualified](https://en.cppreference.com/w/cpp/language/qualified_lookup)) unless a matching declaration is provided at namespace scope, either before or after the class definition. Such name may be found through [ADL](https://en.cppreference.com/w/cpp/language/adl) which considers both namespaces and classes.

So a friend function belongs to the enclosing namespace, not to the class it is friend of. If the function was supposed to be a class member, it would have been declared static and not friend in C++.

In the current global `torch.java`, without this patch, there is already 14 `notEquals`, 5 `equals`, 25 `shiftLeft`, etc... and many other operators overload. So the most consistent seems to put the one defined as friend there too. Also I think your example about paired operators has little chance to happen: why `==` would have been defined as a member function and `!=` as a friend one ?

In fact the best user-friendly, java-ish, solution would be turn all operators overloads as instance methods, be them declared as such, as global not friend functions, or as friend functions. We would need to patch the generator. Is that possible ? One major problem is the case of global not-friend function, since the declaration can appear anywhere, far from the class definition. That would also break some existing code but that may be worth. And concerning non-operator friend functions, we would keep them as static methods (of global, which I think correspond the best to the C++ semantic, or of the friend class. The important is that the method is mapped).

Reply:
> So a friend function belongs to the enclosing namespace, not to the class it is friend of. If the function was supposed to be a class member, it would have been declared static and not friend in C++.

Technically, yes, but like I said, if that function is meant to be used outside the class, it's going to have a static declaration there as well. The reason friends function exist in the first place is because there is no way to declare functions like `operator <<(ostream& os, ...)` as member functions. It's just a hack they came up with. But the intention is usually to treat them like member functions.

> In the current global `torch.java`, without this patch, there is already 14 `notEquals`, 5 `equals`, 25 `shiftLeft`, etc... and many other operators overload. So the most consistent seems to put the one defined as friend there too. Also I think your example about paired operators has little chance to happen: why `==` would have been defined as a member function and `!=` as a friend one ?

If the functions that are friended also have declarations in the global namespace, JavaCPP will continue to map there as well. You haven't made any changes in that respect, right? If the C++ API uses friend functions for some things but not others, that's a problem with original C++ API, not with JavaCPP. Do you have a concrete example that you're worried about?

Why would they declare `operator ==` as a member function and `operator !=` as a friend function indeed. Like I said, that's a problem with the original C++ API. Mapping both of them to `Object.equals()` seems fine though.

> In fact the best user-friendly, java-ish, solution would be turn all operators overloads as instance methods, be them declared as such, as global not friend functions, or as friend functions. We would need to patch the generator. Is that possible ? One major problem is the case of global not-friend function, since the declaration can appear anywhere, far from the class definition. That would also break some existing code but that may be worth. And concerning non-operator friend functions, we would keep them as static methods (of global, which I think correspond the best to the C++ semantic, or of the friend class. The important is that the method is mapped).

Sure, we can do a lot of things, but what I try to do with JavaCPP is not get into twisted situations that don't have a clear mapping from C++. Moving functions around like that just makes it harder for everyone to figure out what is going on. If we want something that's actually designed for Java or Scala, like @sbrunk is doing, we'd have to design manually the whole API from scratch anyway.

Reply:
> Technically, yes, but like I said, if that function is meant to be used outside the class, it's going to have a static declaration there as well.

Not if the function is an operator. Since operator are used with ADL.

Reply:
>  JavaCPP will continue to map there as well. You haven't made any changes in that respect, right? 

I didn't but if we follow your idea, won't we end with 2 mapping for the function? a static function when we encounter the friend declaration, and a global static one when we encounter the declaration outside the class.

Reply:
> Do you have a concrete example that you're worried about?

Let take two examples in Pytorch:

1. `module_iterator`, an instance of the template `torch::jit::slot_iterator_impl` defined in `jit/api/module.h`. `!=` is missing, but `++` is mapped as an instance method in `ModuleIterator.java`.
Because they are defined like this:
```c++
namespace torch {
namespace jit {

template <typename Policy>
struct slot_iterator_impl {
  slot_iterator_impl& operator++() {
    next_valid();
    return *this;
  }
  friend inline bool operator!=(
      const slot_iterator_impl<Policy>& a,
      const slot_iterator_impl<Policy>& b) {
    return (a.cursors_.empty() != b.cursors_.empty());
  } 
}
}
}
```
The developer chose to implement `++` as an instance method and `!=` as a friend method. `!=` is clearly meant to be visible. He could have implemented `!=` as an instance method. Would have been better for us but that wouldn't change anything for the C++ user that writes `i != end` in both cases. No need to declare it in the enclosing namespace too. If he did impement `!=` as an instance method, it would have been mapped as a Java instance method. If he added a declaration in enclosing namespace, it would have been mapped as static in `global.torch`. Never as a static method of `module_iterator`.

2. `TypeIdentifier`, defined in `c10/util/typeid.h`
```c++
class C10_API TypeIdentifier final
    : public at::IdWrapper<TypeIdentifier, c10::util::type_index> {
public:
  friend std::ostream& operator<<(std::ostream& stream, TypeIdentifier typeId);
  friend constexpr bool operator<(TypeIdentifier lhs, TypeIdentifier rhs);
};

// Allow usage in std::map / std::set
inline constexpr bool operator<(TypeIdentifier lhs, TypeIdentifier rhs) {
  return lhs.underlyingId() < rhs.underlyingId();
}
inline std::ostream& operator<<(
    std::ostream& stream,
    caffe2::TypeIdentifier typeId) {
  return stream << typeId.underlyingId();
}

```
`<<` is implemented as a friend function. Like you pointed it out, there was no choice here. 
`<` too, but the reason apparently here is that it needs to be redeclared at the namespace scope in order to be visible by `std::map` when a `TypeIdentifier` is used as a key. `<<` is also redeclared, for good measure I guess.
Both functions are thus mapped in `global.torch`. No need for this PR here.  No static functions in `TypeIdentifier`.

> Why would they declare operator == as a member function and operator != as a friend function indeed. Like I said, that's a problem with the original C++ API. Mapping both of them to Object.equals() seems fine though.

How would you map the friend function to `Object.equals() ` ?

> Sure, we can do a lot of things, but what I try to do with JavaCPP is not get into twisted situations that don't have a clear mapping from C++.

I agree, that's what I think sticking with the C++ semantic is best.

>  like @sbrunk is doing,

I missed that. What is he working on ?

Reply:
> I missed that. What is he working on ?

I'm working on an idiomatic Scala API on top the the PyTorch bindings: https://storch.dev/

Reply:
Cool. Have you seen [this discussion about Pytorch 2](https://discuss.pytorch.org/t/pytorch-2-and-the-c-interface/168034) ?

Reply:
Thanks, that's pretty interesting. The most interesting thing, the answer to your second question remains open though.

Reply:
> Not if the function is an operator. Since operator are used with ADL.

Sure, of course we can. In C++ we can declare the same function any number of times. PyTorch deliberately doesn't define them in global namespace as well because they want us to think of them as sort-of-member function, I guess. So, why not take the hint? It's easier for JavaCPP that way too.

> I didn't but if we follow your idea, won't we end with 2 mapping for the function? a static function when we encounter the friend declaration, and a global static one when we encounter the declaration outside the class.

Sure, that's what it looks like in C++ too, so why is it a problem if we map them 1-to-1 with duplicates? It's the easiest thing to do and it works. If it's not broken, don't fix it!

> How would you map the friend function to Object.equals() ?

Something like I've done for multiple inheritance around here:
https://github.com/bytedeco/javacpp/blob/master/src/main/java/org/bytedeco/javacpp/tools/Parser.java#L3386
That all ends up in Declaration.text. We can *literally* put anything we want there.

Reply:
> > How would you map the friend function to Object.equals() ?
> 
> Something like I've done for multiple inheritance around here: https://github.com/bytedeco/javacpp/blob/master/src/main/java/org/bytedeco/javacpp/tools/Parser.java#L3386 That all ends up in Declaration.text. We can _literally_ put anything we want there.

Ok, so with this trick we could map as instance methods all operators declared as friends and taking an instance of the class or struct as its first argument. Can we agree on that ?

Reply:
> Ok, so with this trick we could map as instance methods all operators declared as friends and taking an instance of the class or struct as its first argument. Can we agree on that ?

Sure, that sounds fine, although apart from equals(), I don't think we gain much by doing so...?

Reply:
Ok, I implemented this: map the friend function to a `private static native` method of the class the function is friend of, and add a Java wrapper as an instance method. It's what you had in mind, right ? 
I had to force the use of ADL instead of namespace lookup with a `@Namespace` annotation without argument or it didn't compile. So now all friend functions accepting at least 1 argument of the class it is friend of are accessible as instance method. 

Example:
```Java
  public boolean notEquals(StringView rhs) { return notEquals(this, rhs); }
  private static native @Namespace @Cast("bool") @Name("operator !=") boolean notEquals(@Const @ByRef StringView lhs, @Const @ByRef StringView rhs);

  public Pointer shiftLeft(Pointer os) { return shiftLeft(os, this); }
  private static native @Namespace @Cast("std::ostream*") @ByRef @Name("operator <<") Pointer shiftLeft(@Cast("std::ostream*") @ByRef Pointer os, @Const @ByRef StringView dt);
```

Do you agree with this solution ? Any remark before I repost a PR ?

Reply:
When a library developer implements an operator, (s)he usually has the choice between implementing it as a friend function or as an instance method. It doesn't change anything for the C++ user.

The good thing about this solution is that, with it, it doesn't change anything for the Java user either.

Reply:
> Do you agree with this solution ? Any remark before I repost a PR ?

Yes, that looks exactly like what I had in mind. You can update this PR if you want, GitHub keeps all the commits either way. Thanks!